File: backend/apps/core/logger.py
"""
Central Logging Utility for Backend

Controlled via .env: ENABLE_LOGS=true

Usage:
    from apps.core.logger import log
    
    log.info("User logged in", user_id=123)
    log.api("POST /bookings", booking_id=456)
    log.error("Database error", error=str(e))
"""

import os
from typing import Any, Dict, Optional
from datetime import datetime


class Logger:
    """Central logger that can be toggled via environment variable"""
    
    def __init__(self):
        # Read from .env file
        self.enabled = os.getenv("ENABLE_LOGS", "false").lower() == "true"
        self.prefix = "[PrepHub Backend]"
        
        if self.enabled:
            print(f"{self.prefix} Logging ENABLED")
    
    def _format_message(self, level: str, message: str, **kwargs) -> str:
        """Format log message with timestamp and kwargs"""
        timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        parts = [f"{self.prefix} [{timestamp}] [{level}] {message}"]
        
        if kwargs:
            parts.append(f" | {kwargs}")
        
        return "".join(parts)
    
    def info(self, message: str, **kwargs):
        """General info log"""
        if not self.enabled:
            return
        print(self._format_message("INFO", message, **kwargs))
    
    def lifecycle(self, component: str, event: str, **kwargs):
        """Lifecycle event log"""
        if not self.enabled:
            return
        print(self._format_message("LIFECYCLE", f"{component}: {event}", **kwargs))
    
    def api(self, message: str, **kwargs):
        """API endpoint log"""
        if not self.enabled:
            return
        print(self._format_message("API", message, **kwargs))
    
    def db(self, message: str, **kwargs):
        """Database operation log"""
        if not self.enabled:
            return
        print(self._format_message("DB", message, **kwargs))
    
    def service(self, message: str, **kwargs):
        """Service layer log"""
        if not self.enabled:
            return
        print(self._format_message("SERVICE", message, **kwargs))
    
    def auth(self, message: str, **kwargs):
        """Authentication log"""
        if not self.enabled:
            return
        print(self._format_message("AUTH", message, **kwargs))
    
    def error(self, message: str, **kwargs):
        """Error log (always shown)"""
        print(self._format_message("ERROR", message, **kwargs))
    
    def warn(self, message: str, **kwargs):
        """Warning log"""
        if not self.enabled:
            return
        print(self._format_message("WARN", message, **kwargs))
    
    def debug(self, message: str, **kwargs):
        """Debug log"""
        if not self.enabled:
            return
        print(self._format_message("DEBUG", message, **kwargs))
    
    def validation(self, message: str, **kwargs):
        """Validation log"""
        if not self.enabled:
            return
        print(self._format_message("VALIDATION", message, **kwargs))


# Create singleton instance
log = Logger()

File: frontend/src/utils/logger.ts
/**
 * Central Logging Utility for Frontend
 * 
 * Usage:
 * - Import: import logger from '@/utils/logger'
 * - Log: logger.info('Component mounted')
 * - Control via .env: VITE_ENABLE_LOGS=true
 */

const ENABLE_LOGS = import.meta.env.VITE_ENABLE_LOGS === 'true';

class Logger {
  private enabled: boolean;
  private prefix: string;

  constructor(enabled = ENABLE_LOGS) {
    this.enabled = enabled;
    this.prefix = '[PrepHub]';
  }

  /**
   * Enable or disable logging globally
   */
  setEnabled(enabled: boolean) {
    this.enabled = enabled;
    console.log(`${this.prefix} Logging ${enabled ? 'ENABLED' : 'DISABLED'}`);
  }

  /**
   * Check if logging is enabled
   */
  isEnabled(): boolean {
    return this.enabled;
  }

  /**
   * General info log
   */
  info(message: string, ...args: any[]) {
    if (!this.enabled) return;
    console.log(`${this.prefix} [INFO]`, message, ...args);
  }

  /**
   * Component lifecycle log
   */
  lifecycle(component: string, event: string, ...args: any[]) {
    if (!this.enabled) return;
    console.log(`${this.prefix} [LIFECYCLE] ${component}:`, event, ...args);
  }

  /**
   * User interaction log
   */
  interaction(action: string, details?: any) {
    if (!this.enabled) return;
    console.log(`${this.prefix} [INTERACTION]`, action, details || '');
  }

  /**
   * API call log
   */
  api(method: string, endpoint: string, payload?: any) {
    if (!this.enabled) return;
    console.log(`${this.prefix} [API] ${method.toUpperCase()}`, endpoint, payload || '');
  }

  /**
   * API response log
   */
  apiResponse(endpoint: string, status: number, data?: any) {
    if (!this.enabled) return;
    console.log(`${this.prefix} [API RESPONSE]`, endpoint, `Status: ${status}`, data || '');
  }

  /**
   * Navigation log
   */
  navigation(from: string, to: string) {
    if (!this.enabled) return;
    console.log(`${this.prefix} [NAVIGATION]`, `${from} → ${to}`);
  }

  /**
   * State change log
   */
  state(component: string, state: string, value: any) {
    if (!this.enabled) return;
    console.log(`${this.prefix} [STATE] ${component}:`, state, value);
  }

  /**
   * Error log (always shown regardless of enabled flag)
   */
  error(message: string, error?: any) {
    console.error(`${this.prefix} [ERROR]`, message, error || '');
  }

  /**
   * Warning log
   */
  warn(message: string, ...args: any[]) {
    if (!this.enabled) return;
    console.warn(`${this.prefix} [WARN]`, message, ...args);
  }

  /**
   * Debug log (for detailed debugging)
   */
  debug(message: string, ...args: any[]) {
    if (!this.enabled) return;
    console.debug(`${this.prefix} [DEBUG]`, message, ...args);
  }

  /**
   * Form interaction log
   */
  form(action: string, formName: string, data?: any) {
    if (!this.enabled) return;
    console.log(`${this.prefix} [FORM] ${action}:`, formName, data || '');
  }

  /**
   * Authentication log
   */
  auth(action: string, details?: any) {
    if (!this.enabled) return;
    console.log(`${this.prefix} [AUTH]`, action, details || '');
  }
}

// Create singleton instance
const logger = new Logger();

// Expose to window for manual control in console
if (typeof window !== 'undefined') {
  (window as any).logger = logger;
}

export default logger;
